staging/vhost-device-gpu/src/device.rs

```diff
@@ -0,0 +1,1607 @@
+                return Err(
+                    Error::EpollHandler("Failed to upgrade epoll handler".to_string()).into(),
+                );
+            };
+            epoll_handler
+                .register_listener(
+                    poll_event_fd.as_raw_fd(),
+                    EventSet::IN,
+                    u64::from(POLL_EVENT),
+                )
+                .map_err(Error::RegisterEpollListener)?;
+            debug!("Registered POLL_EVENT on fd: {}", poll_event_fd.as_raw_fd());
+            // store the fd, so it is not closed after exiting this scope
+            self.poll_event_fd.lock().unwrap().replace(poll_event_fd);
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{
+        fs::File,
+        io::{ErrorKind, Read},
+        iter::zip,
+        mem,
+        os::{fd::FromRawFd, unix::net::UnixStream},
+        sync::Arc,
+        thread,
+        time::Duration,
+    };
+
+    use assert_matches::assert_matches;
+    use mockall::predicate;
+    use rusty_fork::rusty_fork_test;
+    use tempfile::tempdir;
+    use vhost::vhost_user::gpu_message::{VhostUserGpuScanout, VhostUserGpuUpdate};
+    use vhost_user_backend::{VhostUserDaemon, VringRwLock, VringT};
+    use virtio_bindings::virtio_ring::{VRING_DESC_F_NEXT, VRING_DESC_F_WRITE};
+    use virtio_queue::{mock::MockSplitQueue, Descriptor, Queue, QueueT};
+    use vm_memory::{
+        ByteValued, Bytes, GuestAddress, GuestMemory, GuestMemoryAtomic, GuestMemoryMmap,
+    };
+
+    use super::*;
+    use crate::{
+        protocol::{
+            virtio_gpu_mem_entry, virtio_gpu_rect, virtio_gpu_resource_attach_backing,
+            virtio_gpu_resource_flush, virtio_gpu_set_scanout,
+            GpuResponse::{OkCapsetInfo, OkDisplayInfo, OkEdid, OkNoData},
+            VIRTIO_GPU_CMD_RESOURCE_ATTACH_BACKING, VIRTIO_GPU_CMD_RESOURCE_CREATE_2D,
+            VIRTIO_GPU_CMD_RESOURCE_FLUSH, VIRTIO_GPU_CMD_SET_SCANOUT,
+            VIRTIO_GPU_CMD_TRANSFER_FROM_HOST_3D, VIRTIO_GPU_CMD_TRANSFER_TO_HOST_2D,
+            VIRTIO_GPU_CMD_TRANSFER_TO_HOST_3D, VIRTIO_GPU_FORMAT_R8G8B8A8_UNORM,
+            VIRTIO_GPU_RESP_ERR_UNSPEC, VIRTIO_GPU_RESP_OK_NODATA,
+        },
+        virtio_gpu::MockVirtioGpu,
+    };
+
+    const SOCKET_PATH: &str = "/tmp/vgpu.socket";
+    const MEM_SIZE: usize = 2 * 1024 * 1024; // 2MiB
+
+    const CURSOR_QUEUE_ADDR: GuestAddress = GuestAddress(0x0);
+    const CURSOR_QUEUE_DATA_ADDR: GuestAddress = GuestAddress(0x1_000);
+    const CURSOR_QUEUE_SIZE: u16 = 16;
+    const CONTROL_QUEUE_ADDR: GuestAddress = GuestAddress(0x2_000);
+    const CONTROL_QUEUE_DATA_ADDR: GuestAddress = GuestAddress(0x10_000);
+    const CONTROL_QUEUE_SIZE: u16 = 1024;
+
+    fn init() -> (Arc<VhostUserGpuBackend>, GuestMemoryAtomic<GuestMemoryMmap>) {
+        let test_dir = tempdir().expect("Could not create a temp test directory.");
+        let socket_path = test_dir.path().join(SOCKET_PATH);
```

Comment:
`SOCKET_PATH` should be just the filename here, because `/tmp/[..]` is absolute and `Path::join` replaces the path with the argument if it's absolute (source <https://doc.rust-lang.org/std/path/struct.Path.html#method.join>). It's a bad gotcha, this method. So doing `test_dir.path().join("/tmp/vgpu.socket")` is the same as `Path::new("/tmp/vgpu.socket")`.

```suggestion
    const SOCKET_PATH: &str = "vgpu.socket";
    const MEM_SIZE: usize = 2 * 1024 * 1024; // 2MiB

    const CURSOR_QUEUE_ADDR: GuestAddress = GuestAddress(0x0);
    const CURSOR_QUEUE_DATA_ADDR: GuestAddress = GuestAddress(0x1_000);
    const CURSOR_QUEUE_SIZE: u16 = 16;
    const CONTROL_QUEUE_ADDR: GuestAddress = GuestAddress(0x2_000);
    const CONTROL_QUEUE_DATA_ADDR: GuestAddress = GuestAddress(0x10_000);
    const CONTROL_QUEUE_SIZE: u16 = 1024;

    fn init() -> (Arc<VhostUserGpuBackend>, GuestMemoryAtomic<GuestMemoryMmap>) {
        let test_dir = tempdir().expect("Could not create a temp test directory.");
        let socket_path = test_dir.path().join(SOCKET_PATH);
```

Reply:
I’ll add your suggestion to just use the filename instead of an absolute path. That way, it behaves as intended and properly joins with `test_dir.path()`. Thanks!